### PR TITLE
Paul Haddick

### PR DESCRIPTION
**Pull request by phdd0**
My completed M1 Prework...

### CODE REVIEW

File: final_prep/mod_zero_hero.rb [hunk 1]
```diff
@@ -1,84 +1,115 @@
...
+energy = 100
+full_power = power * 500
+full_energy = energy + 150
+is_human = false
+identity_concealed = true
+arch_enemies = ["GeorgeTheToadTerror", "RandallTheMangyMantis", "EVILSloth"]
+sidekicks = ["CricketCraig", "AntAndy", "RoachRalph"]
+
+puts "#{sidekicks[0]}"
+puts "#{arch_enemies[2]}"
```
Comment (philmccarthy) [inline final_prep/mod_zero_hero.rb:17]: Interpolation is redundant here. `puts sidekicks[0]` alone would return a string

File: final_prep/mod_zero_hero.rb [hunk 2]
```diff
@@ -1,84 +1,115 @@
...
+puts "assess_situation(51, save_the_day, bad_excuse)"
+assess_situation(51, save_the_day, bad_excuse)
+puts "The following method call has first parameter 50 (~> 50) so should be 'Darlings':"
+puts "assess_situation(50, save_the_day, bad_excuse)"
+assess_situation(50, save_the_day, bad_excuse)
+puts "The following method call has first parameter 9 (~> 9) so should be 'Meh':"
+puts "assess_situation(9, save_the_day, bad_excuse)"
+assess_situation(9, save_the_day, bad_excuse)
+
+scary_monster = { "name" => "Heffalump", "smell" => "Atrocious", "weight" => 1000000, "cities_destroyed" => ["carthage", "babel", "memphis"], "lucky_numbers" => [7, 5, 3, 9, 13, 119], "address" => { "number" => 221, "street" => "Long Windy Lane", "state" => "CO", "zip" => "88888" }}
```
Comment (philmccarthy) [inline final_prep/mod_zero_hero.rb:52]: Please work on formatting hashes with line breaks and indentation. A long, one-line hash isn't very kind to other developers who read your code 😉

File: final_prep/mod_zero_hero.rb [hunk 3]
```diff
@@ -1,84 +1,115 @@
...
+
+class SuperHero
+  def initialize(name, super_power, age)
+    @@name = name
+    @@super_power = super_power
+    @@age = age
```
Comment (philmccarthy) [inline final_prep/mod_zero_hero.rb:58]: Single `@` for instance variables. You'll learn more about the `@@` class variable syntax in Mod 1, but feel free to do some research into why class variables aren't appropriate in this example! You'll also find some good commentary on why class variables can lead to unintended behavior in Ruby.

But word to the wise: you shouldn't _need_ to use class variables _at all_ in Mod 1.

File: final_prep/mod_zero_hero.rb [hunk 4]
```diff
@@ -1,84 +1,115 @@
...
+    # The hardest part was figuring out where outputs should go given where the variables are
+    # and how they behave, but hey now I really get the static vs. dynamic part: The class "@@"
+    # variables are "dynamic" since they are updatable as arguments every time .new >> .initialize
+    # gets called and set for the class instance BY those arguments. The instance or object "@" 
+    # variables only apply to the specific instance created, like 'sam' and are not updatable by 
+    # argument, each instantiation of the class gets power and energy levels of 100 and 50 
+    # respectively once spawned but the name, super power and age dynamic class variables are 
+    # assigned by argument when calling Class.new
+    # Really then the only 'hard' part was really figuring out how to make the methods work with
+    # variable referencing
```
Comment (philmccarthy) [inline final_prep/mod_zero_hero.rb:104]: You may have conflated the exercise's `dynamic` reference with affecting all instances of a class, I'm not exactly following where you took this. 

In this exercise, a dynamic attribute should only be dynamic for the instance of the class you are creating. Both dynamic and static attributes for an instance of SuperHero would be assigned as instance variables (`@` not `@@`).

File: final_prep/annotations.rb [hunk 5]
```diff
@@ -18,11 +23,29 @@ def build_a_bear(name, age, fur, clothes, special_power)
...
+# input integers (though if someone types in some decimals causing Ruby to store all
+# the arguments as floats the result will nearly always return a non 0 modulus)
+# and then starts a loop to first check to see whether the range index divided by 
+# the first arg has modulus or remainder identically 0 AND whether this is also 
+# true for the second arg provided at the sametime, resulting in 'fizzbuzz' on the 
+# screen only in that case, this for example occurs say when range is 1 to 3 when 
+# args are (1, 1, range); if the test fails then Ruby checks whether it was ONLY
+# the first arg OR (exclusively) the second that has zero remainder printing
+# corresponding output in those cases, in case no test returns true Ruby will just
+# print out the range index
```
Comment (philmccarthy) [inline final_prep/annotations.rb:42]: the else statement prints `i` (a number from the range, not index) if the previous conditionals are false

You took this in some interesting directions though, love that you're thinking about edge cases so early. But in this exercise, I would've liked to see you focus on the test cases as they were written instead of changing them to floats.

File: final_prep/annotations.rb [hunk 6]
```diff
@@ -18,11 +23,29 @@ def build_a_bear(name, age, fur, clothes, special_power)
...
+
+# this is a fairly decent random string sequence generator as long as the person typing
+# in the numbers doesn't think about what it's doing
```
Comment (philmccarthy) [inline final_prep/annotations.rb:45]: do some googling of fizzbuzz - it's a common problem you might see in technical interviews. It's less about random string generation, more about pattern matching

File: final_prep/mod_zero_hero.rb [hunk 7]
```diff
@@ -1,84 +1,115 @@
...
+puts "The sidekicks are: #{sidekicks} but wait!..."
+sidekicks.shift
+puts "Now the sidekicks are #{sidekicks}..."
+
+save_the_day = "ISaveYouGoodlyDarlings!"
+bad_excuse = "Sorry, HOT, gotta pee!"
+
+def assess_situation(danger_level, save_the_day, bad_excuse)
+  if danger_level > 50
+    puts "#{bad_excuse}"
```
Comment (philmccarthy) [inline final_prep/mod_zero_hero.rb:34]: Similar to above, no need to interpolate a single string variable. `puts bad_excuse` would do the trick

File: final_prep/annotations.rb [hunk 8]
```diff
@@ -4,6 +4,11 @@
 
 # Build a Bear
 
+# method build_a_bear takes input args, outputs two to named string variables 
+# and outputs two to demographic array and mixes hard defined info common
+# to all build_a_bear method buildable bears into a hash with these strings
+# and variable stored information
+
```
Comment (philmccarthy) [inline final_prep/annotations.rb:11]: nice job 👍🏻

File: final_prep/README.md [hunk 9]
```diff
@@ -19,37 +19,37 @@ Some things to consider as you refactor include...
...
+- [x] Health + Wellness
 
 When you are finished, add screenshots of your calendar so we can provide feedback if needed!
 
-- `Add Week 1 Screenshot Here`
-- `Add Week 2 Screenshot Here`
-- `Add Week 3 Screenshot Here`
+- ![wk1](https://user-images.githubusercontent.com/94933767/146654439-3ccadb3d-57f1-4501-89b4-8bcd017aa52b.JPG)
+- ![wk2](https://user-images.githubusercontent.com/94933767/146654456-903931fe-c0fe-4563-b56b-d0fa4ddd31db.JPG)
+- ![wk3](https://user-images.githubusercontent.com/94933767/146654463-630a8507-6564-45c4-b82c-ecacfd7c3b23.JPG)
```
Comment (philmccarthy) [inline final_prep/README.md:37]: Calendar looks good!

File: final_prep/README.md [hunk 10]
```diff
@@ -19,37 +19,37 @@ Some things to consider as you refactor include...
...
 
 ### Mentorship Prep
 Mentorship is an integral part of the Turing experience and will help jumpstart your technical career. In order to get your mentor relationship started on the right foot, please complete the following deliverables:
-- [ ] Complete the [Mentorship DTR Prep](https://gist.github.com/ericweissman/51965bdcbf42970d43d817818bfaef3c)
-  - [ ] Add link to your gist here: 
+- [x] Complete the [Mentorship DTR Prep](https://gist.github.com/ericweissman/51965bdcbf42970d43d817818bfaef3c)
+  - [x] Add link to your gist here: [Link to my Mentorship DTR Prep](https://gist.github.com/phdd0/5124f36bcbe74bcce3128901c295e532)
```
Comment (philmccarthy) [inline final_prep/README.md:42]: Your draft to a potential mentor looks good! Maybe end it with a question for them instead of leaving it kinda open ended? Totally your call on how best to 'sell' yourself when the time comes though.

File: final_prep/README.md [hunk 11]
```diff
@@ -19,37 +19,37 @@ Some things to consider as you refactor include...
...
 
 ### Lesson Prep
 You've learned a lot about how to take strong notes during Mod 0. Show us your skills while you learn how to pre-teach content for your first lesson in Mod 1!
-- [ ] Complete the [Pre Teaching Practice exercise](https://gist.github.com/ericweissman/0036e8fe272c02bd6d4bb14f42fd2f79) gist
-  - [ ] Add a link to your gist here:
+- [x] Complete the [Pre Teaching Practice exercise](https://gist.github.com/ericweissman/0036e8fe272c02bd6d4bb14f42fd2f79) gist
+  - [x] Add a link to your gist here:[Link to my Pre Teaching Prac](https://gist.github.com/phdd0/82889033e2f8c55a728f09bdf1d283f8)
```
Comment (philmccarthy) [inline final_prep/README.md:47]: Keep up the great work on notetaking.

File: final_prep/README.md [hunk 12]
```diff
@@ -19,37 +19,37 @@ Some things to consider as you refactor include...
...
 
 ### Group Work Prep
 As part of Turing's project-based learning approach, you will often be working in pairs or larger groups. In order to set yourself (and your team) up for success, it is important to ensure you are prepared to be an equitable contributor and teammate.
-- [ ] Complete the [DTR Guiding Questions](https://gist.github.com/ericweissman/c56f3a98cdce761808c21d498a52f5c6)
-  - [ ] Add a link to your gist here: 
+- [x] Complete the [DTR Guiding Questions](https://gist.github.com/ericweissman/c56f3a98cdce761808c21d498a52f5c6)
+  - [x] Add a link to your gist here: [Link to my DTR Guiding Question Answers](https://gist.github.com/phdd0/b50926a18f00e906309899bcbe23f0ec)
```
Comment (philmccarthy) [inline final_prep/README.md:52]: Lots of thoughtful reflections in here. Stay humble and help your teammates shine 👍🏻

File: section4/exercises/burrito.rb [hunk 13]
```diff
@@ -5,15 +5,31 @@
...
+  
   attr_reader :protein, :base, :toppings
+  
   def initialize(protein, base, toppings)
     @protein  = protein
     @base     = base
     @toppings = toppings
   end
+
+  def add_topping()
```
Comment (philmccarthy) [inline section4/exercises/burrito.rb:17]: I see some interesting stuff going on in your `_broaden` file. I would recommend spending some time looking into the Mod 0 extensions, particularly Ruby Exercises > Mythical Creatures to get some practice with these class concepts.

The biggest learning opportunity I (maybe?) see is to get away from thinking about screen output, and instead about an object's state.

You'll get a lot more exposure to classes and sort this out for yourself in Mod 1, as you suggested in your reflection. 

Throwing some code in here for you to ponder another approach to writing these methods without print statements.

```rb
def add_topping(new_topping)
  # Add new_topping to burrito's toppings array (<< is an array method)
  @toppings << new_topping
end

def remove_topping(topping)
  # Remove topping from burrito's toppings array (delete is an array method)
  @toppings.delete(topping)
end

def change_protein(new_protein)
  # Reassign the burrito's @protein attribute
  @protein = new_protein
end
```

File: section4/exercises/burrito.rb [hunk 14]
```diff
@@ -5,15 +5,31 @@
...
+  def change_protein()
+  end
+
 end
 
 dinner = Burrito.new("Beans", "Rice", ["cheese", "salsa", "guacamole"])
 p dinner.protein
 p dinner.base
 p dinner.toppings
+
```
Comment (philmccarthy) [inline section4/exercises/burrito.rb:32]: Given above example, these three method calls would each have 1 argument.

Example of how the `dinner` instance of Burrito's state changes as we interact with the object using its `instance methods`... 

```rb
dinner.add_topping("queso")

dinner.toppings
  # => ["cheese", "salsa", "guacamole", "queso"]

dinner.remove_topping("cheese")
  # => "cheese"

dinner.toppings
  # => ["salsa", "guacamole", "queso"]

dinner.remove_topping("non existent topping")
  # => nil

dinner.toppings
  # => ["salsa", "guacamole", "queso"]

dinner.change_protein("Chicken")
 
dinner.protein
  # => "Chicken"
```

These methods each have return values and `dinner`'s state is changing even though we don't print to the screen in the above interaction.

File: section4/exercises/burrito_broaden.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    puts "this is the new toppings: #{toppings}" + "and of course previous toppings were #{toppings_prev_add}"
+
+    #toppings_prev0 = ["cheese", "salsa", "guacamole"]
+    #toppings = ["cheese", "salsa", "guacamole", "sour #cream"]
+    #puts "Now you have a " + "#{protein}" + ", rice #burrito with " + "#{toppings_prev0}" + " but we've #added " + "#{toppings[3]}" + "\n\n"
+  end
+  
+  def remove_topping
+    toppings_prev1 = toppings
+    toppings -= ["cheese", "guacamole"]
```
Comment (philmccarthy) [inline section4/exercises/burrito_broaden.rb:35]: would need to be `@toppings` to change the attribute

File: section4/exercises/burrito_broaden.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    puts "Everyone knows a #{protein}, rice burrito with ONLY #{toppings} and lacking #{toppings_prev1[0]} and #{toppings_prev_1[2]} as we've removed both, are pretty lame and boring..."
+  end
+
+  def change_protein
+    protein_prev = protein
+    protein = "Beef"
```
Comment (philmccarthy) [inline section4/exercises/burrito_broaden.rb:41]: would need to be `@protein` to change the attribute

File: section4/exercises/burrito_broaden.rb [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    @toppings = toppings
+  end
+ 
+  ### first pass method concept only, need to add coordination of these 
+  ### operations together to yield correct results for screen output
+  ### this code will be sufficient for testing each opertion only
+  ### variable scheme TBD to optimally model burrito changes
+
+  def add_topping
+    toppings = toppings_prev_add
```
Comment (philmccarthy) [inline section4/exercises/burrito_broaden.rb:23]:
```
NameError (undefined local variable or method `toppings_prev_add' for #<Burrito:0x00007fdb2a975ec0>)
```

assignment is backwards

File: section4/exercises/burrito_broaden.rb [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    puts "this is toppings now:" + "#{toppings}" + " and this is the previous toppings: #{toppings_prev_add} in a new variable"
+    toppings += ["sour cream"]
```
Comment (philmccarthy) [inline section4/exercises/burrito_broaden.rb:25]: would need to be @toppings to change the attribute

File: section4/exercises/dog.rb [hunk 19]
```diff
@@ -19,6 +19,13 @@ def bark
   def eat
     @hungry = false
   end
+  
+  def play
+    @hungry = true
```
Comment (philmccarthy) [inline section4/exercises/dog.rb:24]: Nice—here we're changing the `@hungry` instance variable (attribute). 

Good example to compare to methods in the Burrito exercise that aren't altering instance variables

File: section4/reflection.md [hunk 20]
```diff
@@ -2,21 +2,55 @@
...
+    * Select a variable name to store the instance in, then set it equal to YourClass.new
+    * This creates an instance of the class with the selected name
+
+8. What questions do you still have about classes in Ruby?
 
-1. What questions do you still have about classes in Ruby?
\ No newline at end of file
+    * As in my burrito_broaden.rb file I'm not sure about results I was getting attempting to do things inside class (instance) methods
+    * I'm still not clear at all at how the syntax in particular conveys the ideas expressed in the technical language about what the code is doing
+    * If someone were to walk me thorough how all of this particular syntax expresses the underlying structure of class and variables and methods accessible at various levels I imagine I'd understand more clearly
```
Comment (philmccarthy) [inline section4/reflection.md:55]: Hopefully my explanations help out a bit with this. You're doing great, and you're right on track with figuring this stuff out, and I appreciate the amount of effort you're putting in. Keep up the great work in Mod 1 🙂

**Review by philmccarthy — COMMENTED** [review]
Hey @phdd0, I just had a chance to go over your prework and you are officially **Turing Ready**!

I added some comments to a few sections for areas to work on or adjustments to make before starting Mod 1 and put a little extra into your class exercises. Hope this helps bring some clarity, but you're doing great and a lot of this will become clearer as you get into Mod 1.

Keep up the hard work moving forward and let me know if you have any questions.
